astropy/units/format/utils.py

```diff
@@ -4,13 +4,27 @@
 Utilities shared by the different formats.
 """
 
+from __future__ import annotations
+
 import warnings
+from typing import TYPE_CHECKING
 
 from astropy.units.utils import maybe_simple_fraction
 from astropy.utils.misc import did_you_mean
 
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterable, Mapping, Sequence
+    from numbers import Real
+    from typing import TypeVar
+
+    from astropy.units import UnitBase
+
+    UnitT = TypeVar("UnitT", bound=UnitBase)
+
 
-def get_grouped_by_powers(bases, powers):
+def get_grouped_by_powers(
+    bases: Sequence[UnitT], powers: Sequence[int]
```

Comment:
This annotation differs from the docstrings, which should be updated.
Also, can't `bases` and `powers` be any `Iterable[...]`, not just a `Sequence`?

Reply:
We need a type that would forbid e.g. `set` instances because `set` makes no promises about iteration order. From what I can tell `bases` and `powers` are always `list` instances in practice, so `Sequence` should be fine.

Reply:
Good enough!